dali/pipeline/operators/arithmetic/expression_impl_factory_gpu.cu

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "dali/pipeline/operators/arithmetic/expression_tree.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_gpu.cuh"
+#include "dali/pipeline/operators/arithmetic/expression_impl_factory.h"
+
+// float16
+#define ALLOWED_TYPES \
+  (uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, float, double)
+
+#define ALLOWED_OPS \
+  (ArithmeticOp::add, ArithmeticOp::sub, ArithmeticOp::mul, ArithmeticOp::div, ArithmeticOp::mod)
+
+namespace dali {
+
+std::unique_ptr<ExpressionImplBase> ExprImplFactory(const DeviceWorkspace &ws,
```

Comment:
Isn't that the same as `std::unique_ptr<ExpressionImplBase> ExprImplFactory(const HostWorkspace &ws, const ExprNode &expr) {` but only ExpressionImplBinGPU<->ExpressionImplBinCPU

Reply:
Yup, I can templetize it for now.

Reply:
+1